### PR TITLE
Feat/validate

### PR DESCRIPTION
**Pull request by greged93**
Add the validation entrypoint call in the execution pipeline.

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [x] Testing
- [ ] Other (please describe):

# What is the current behavior?

Currently, the `__validate__` entrypoint is not called on the account contract executing the transaction.

Issue Number: #227 

# What is the new behavior?

- Add `validate_tx` in order to call the correct validation entrypoint on the sender account
- Add general testing with various validation patterns (Argent, Braavos, Openzeppelin)
- Add edge cases to the test suite

# Does this introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: crates/pallets/starknet/src/tests/declare_tx.rs [hunk 1]
```diff
@@ -119,3 +98,58 @@ fn given_contract_declare_tx_fails_wrong_tx_version() {
...
+            sender_address: account_addr,
+            contract_class: erc20_class,
+            version: 1,
+            compiled_class_hash: erc20_class_hash,
+            nonce: U256::zero(),
+            max_fee: U256::from(u128::MAX),
+            signature: sign_message_hash(tx_hash),
+        };
+
+        assert_ok!(Starknet::declare(none_origin, transaction));
```
Comment (LucasLvy) [inline crates/pallets/starknet/src/tests/declare_tx.rs:126]: also check that the class is indeed declared and the declared class = the one that you declared. Something like that
```rs
assert_eq!(Starknet::contract_class_from_class_hash(erc20_class_hash), erc20_class)
```

File: crates/primitives/starknet/Cargo.toml [hunk 2]
```diff
@@ -30,15 +30,16 @@ starknet_api = { workspace = true, default-features = false }
 cairo-vm = { workspace = true }
 
 # Other third party dependencies
+bitvec = { workspace = true, features = ["alloc"] }
+hex = { version = "0.4.3", default-features = false }
+itertools = { workspace = true }
```
Comment (LucasLvy) [inline crates/primitives/starknet/Cargo.toml:35]: Why is it needed ?

File: Cargo.toml [hunk 3]
```diff
@@ -131,6 +131,7 @@ scale-info = { version = "2.6.0", default-features = false }
 lazy_static = { version = "1.4.0", default-features = false }
 log = { version = "0.4.17", default-features = false }
 hex = { version = "0.4.3", default-features = false }
+itertools = { version = "0.10.3", default-features = false }
```
Comment (LucasLvy) [inline Cargo.toml:134]: why do you need it ?

**Review by LucasLvy — CHANGES_REQUESTED** [review]
Also compile the contracts with 0.10.3 plz

File: crates/primitives/starknet/src/transaction/mod.rs [hunk 4]
```diff
@@ -298,6 +306,96 @@ impl Transaction {
...
+    pub fn validate_entrypoint_calldata(&self, tx_type: &TxType) -> TransactionExecutionResultWrapper<Calldata> {
+        match tx_type {
+            TxType::Declare => {
+                let declare_tx: DeclareTransaction =
+                    self.try_into().map_err(TransactionExecutionErrorWrapper::StarknetApi)?;
+                Ok(calldata![declare_tx.class_hash().0])
+            }
+            TxType::DeployAccount => {
+                let deploy_account_tx: DeployAccountTransaction =
+                    self.try_into().map_err(TransactionExecutionErrorWrapper::StarknetApi)?;
```
[the remark this answers is not in the document]
Comment (greged93) [reply crates/primitives/starknet/src/transaction/mod.rs:339]: This is the type of error returned by the `try_into`, see [here](https://github.com/keep-starknet-strange/madara/blob/main/crates/primitives/starknet/src/transaction/mod.rs#L178).

File: crates/primitives/starknet/src/transaction/mod.rs [hunk 5]
```diff
@@ -413,11 +519,14 @@ impl Transaction {
...
-                    tx.run_execute(state, execution_resources, &block_context, &account_context, contract_class)
-                        .map_err(TransactionExecutionErrorWrapper::TransactionExecution)?,
-                    account_context,
-                )
+                let transaction_execution = tx
+                    .run_execute(state, execution_resources, &block_context, &account_context, contract_class)
+                    .map_err(TransactionExecutionErrorWrapper::TransactionExecution)?;
+
+                // Validate.
+                self.validate_tx(state, execution_resources, &block_context, &account_context, &tx_type)?;
```
[the remark this answers is not in the document]
Comment (greged93) [reply crates/primitives/starknet/src/transaction/mod.rs:527]: In the case of account deployment, validation should come after execution, see [here](https://github.com/tdelabro/blockifier/blob/main/crates/blockifier/src/transaction/account_transaction.rs#L294).

**Review by abdelhamidbakhta — APPROVED** [review]
lgtm
